aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h

```diff
@@ -0,0 +1,113 @@
+    }
+}
+
+namespace detail {
+
+// with_scattered_tensor_options takes a function pointer that potentially takes a TensorOptions argument.
+// If it does, then it creates a new function pointer that takes scattered arguments, internally
+// gathers those arguments, and then calls the underlying function pointer. If the underlying
+// function pointer does not take a TensorOptions argument, it is passed through unmodified.
+
+template<class Type, class Enable = void> struct is_tensoroptions_arg : std::false_type {};
+template<class Type> struct is_tensoroptions_arg<Type, std::enable_if_t<std::is_same<TensorOptions, std::decay_t<Type>>::value>> : std::true_type {};
+template<class Type>
+using is_tensoroptions_arg_t = typename is_tensoroptions_arg<Type>::type;
+
+template<class FuncType>
+inline constexpr bool has_tensoroptions_arg() {
+    using parameter_types = typename guts::infer_function_traits_t<FuncType>::parameter_types;
+    constexpr size_t num_tensoroptions_args = guts::typelist::count_if<is_tensoroptions_arg_t, parameter_types>::value;
+    static_assert(num_tensoroptions_args <= 1, "Function has multiple TensorOptions parameters. We support at most one.");
+    return num_tensoroptions_args > 0;
+}
+
+// sanity checks
+static_assert(has_tensoroptions_arg<int (int64_t, const TensorOptions&)>(), "");
+static_assert(has_tensoroptions_arg<int (int64_t, TensorOptions)>(), "");
+static_assert(!has_tensoroptions_arg<int (int64_t, std::string)>(), "");
+
+template<class FuncPtr, class ParametersBeforeTensorOptions, class ParametersAfterTensorOptions> struct with_scattered_tensor_options_;
+
+template<class FuncPtr, class Enable = void>
+struct with_scattered_tensor_options final {};
+
+template<class UnderlyingFuncPtr>
+struct with_scattered_tensor_options<UnderlyingFuncPtr, std::enable_if_t<!has_tensoroptions_arg<typename UnderlyingFuncPtr::FuncType>()>> final {
+    // FuncType does not have TensorOptions arguments.
+    // Don't wrap anything but just return the base pointer.
+    using FuncPtr = UnderlyingFuncPtr;
+};
+
+template<class UnderlyingFuncPtr>
+struct with_scattered_tensor_options<UnderlyingFuncPtr, std::enable_if_t<has_tensoroptions_arg<typename UnderlyingFuncPtr::FuncType>()>> final {
+private:
+    // FuncType has TensorOptions arguments.
+    // Return a function pointer to a wrapper function that replaces those with expanded arguments.
+    using gathered_parameter_types = typename guts::infer_function_traits_t<typename UnderlyingFuncPtr::FuncType>::parameter_types;
+    static constexpr size_t tensoroptions_arg_index =
+        guts::typelist::find_if<
+            gathered_parameter_types,
+            is_tensoroptions_arg_t
+        >::value;
+
+    using parameters_before_tensoroptions =
+        guts::typelist::take_t<gathered_parameter_types, tensoroptions_arg_index>;
+    using parameters_after_tensoroptions =
+        guts::typelist::drop_t<gathered_parameter_types, tensoroptions_arg_index + 1>;
```

Comment:
Not gonna lie: I'm impressed this all works at all! Do you remember if there were any hidden bombshells, e.g., compiler bugs, you had to work around to get this working? (What are the buried skeletons?)